docs/proposals/rollout/rollout.md

```diff
@@ -0,0 +1,531 @@
+
+    // Defaults to the RolloutConfig.ServiceName
+    // +optional
+    Hosts []string `json:"hosts,omitempty"`
+
+    // Match conditions of HTTP header.
+    // +optional
+    Match []istiov1alpha3.HTTPMatchRequest `json:"match,omitempty"`
+
+    // Retries policy for Http links.
+    // +optional
+    Retries *istiov1alpha3.HTTPRetry `json:"retries,omitempty"`
+
+    // Headers operations for the Request.
+    // e.g.
+    // headers:
+    //   request:
+    //     add:
+    //       x-some-header: "value"
+    // +optional
+    Headers *istiov1alpha3.Headers `json:"headers,omitempty"`
+
+    // Cross-Origin Resource Sharing policy for the Request.
+    // corsPolicy:
+    //   allowHeaders:
+    //   - x-some-header
+    //   allowMethods:
+    //   - GET
+    //   allowOrigin:
+    //   - example.com
+    //   maxAge: 24h
+    // +optional
+    CorsPolicy *istiov1alpha3.CorsPolicy `json:"corsPolicy,omitempty"`
+
+    // CanaryStrategy defines parameters for canary test.
+    CanaryStrategy CanaryConfig `json:"canaryStrategy,omitempty"`
+}
+
+type CanaryConfig struct {
+    // Max traffic weight routed to canary test
+    // +optional
+    MaxWeight int `json:"maxWeight,omitempty"`
+
+    // StepWeight defines the incremental traffic weight step for analysis phase
+    // If set stepWeight: 10 and set maxWeight: 50
+    // The flow ratio between PREVIEW and PRIMARY at each step is
+    // (10:90) (20:80) (30:70) (40:60) (50:50)
+    // +optional
+    StepWeight int `json:"stepWeight,omitempty"`
+
+    // StepWeights defines the incremental traffic weight steps for analysis phase
+    // Note: Cannot configure StepWeights and StepWeight at the same time.
+    // If both StepWeights and MaxWeight are configured, the traffic 
+    // will be scaled according to the settings in StepWeights only.
+    // If set stepWeights: [1, 10, 20, 80]
+    // The flow ratio between PREVIEW and PRIMARY at each step is
+    // (1:99) (10:90) (20:80) (80:20)
+    // +optional
+    StepWeights []int `json:"stepWeights,omitempty"`
+
+    // StepWeightPromotion defines the incremental traffic weight step for promotion phase
+    // If maxWeight: 50 and set StepWeightPromotion: 20
+    // After a successful test, traffic to the PRIMARY version changes as follows: 50 70 90 100.
+    // +optional
+    StepWeightPromotion int `json:"stepWeightPromotion,omitempty"`
+}
+```
+
+As part of the TrafficAnalysis process, Kurator can validate service level objectives (SLOs) like availability, error rate percentage, average response time and any other objective based on app specific metrics. If a drop in performance is noticed during the SLOs analysis, the release will be automatically rolled back with minimum impact to end-users.
+
+```console
+// Note: refer to https://github.com/fluxcd/flagger/blob/main/pkg/apis/flagger/v1beta1/canary.go
+type TrafficAnalysis struct {
+    // Schedule interval for this traffic analysis
+    Interval string `json:"interval"`
```

Comment:
same as above `Timeout`

Reply:
No. `Interval` is not as same as `Timeout`. Interval is the time interval between each test. Kurator changes the traffic distribution rules (if they need to be changed) and performs a traffic analysis every so often.

Reply:
I mean the time type

Reply:
Time type is same as above `Timeout`